--- ./main.php	2014-11-16 14:53:38.235094774 +0200
+++ ./index.php	2014-11-16 14:52:37.000000000 +0200
@@ -15,6 +15,20 @@
 
 <?php
 
+function setup_session(&$cleartext_pw) {
+	$_SESSION['logged_in'] = 1;
+	$_SESSION['key'] = md5("%dJ9&".strtolower($cleartext_pw)."(/&k.=".strtoupper($cleartext_pw)."1x&%");
+	// delete cleartext pw in memory
+	$cleartext_pw = "";
+	unset($cleartext_pw);
+}
+
+// use master password if HTTP Basic Auth is set
+$authenticated = !empty($_SERVER['PHP_AUTH_USER']) && !empty($_SERVER['PHP_AUTH_PW']);
+if ($authenticated && isset($master_password)) {
+	setup_session($master_password);
+}
+
 	// session active?
 	if (!isset($_SESSION['logged_in']))
 	{
@@ -38,8 +52,9 @@
 				if ($crypt_pw == $db_pw)
 				{
 					// password match - proceed
-					$_SESSION['logged_in'] = 1;
-					$_SESSION['key'] = md5("%dJ9&".strtolower($cleartext_pw)."(/&k.=".strtoupper($cleartext_pw)."1x&%");
+
+					setup_session($cleartext_pw);
+
 					// delete cleartext pw in memory
 					unset($cleartext_pw);
 					$_SESSION['version']=$entries->version;
--- w3pw-1.40/include/config.php	2014-11-16 14:55:26.126783112 +0200
+++ w3pw-1.40/include/config.php	2014-11-16 14:40:39.000000000 +0200
@@ -22,6 +22,9 @@
 // generated when adding a new entry
 $random_pw_length = 12;
 
+// master password -- used when $_SERVER PHP_AUTH_USER and PHP_AUTH_PW are set
+#$master_password = "secret";
+
 // don't report Database-Errors on Frontend
 //error_reporting(E_ERROR);
 ?>
